scheduler/ipp.c

```diff
@@ -5544,20 +5544,55 @@ create_local_printer(
 
  /*
   * Check device URI if it has the same hostname as we have, if so, replace
-  * the hostname by local host. This way we assure that local-only services
+  * the hostname by localhost. This way we assure that local-only services
   * like ipp-usb or Printer Applications always work.
+  *
+  * When comparing our hostname with the one in the device URI, consider
+  * names with ¨.local(.)" suffix and names without suffix the same.
   */
 
-  httpSeparateURI(HTTP_URI_CODING_ALL, ptr,
-		  scheme, sizeof(scheme), userpass, sizeof(userpass), host,
-		  sizeof(host), &port, resource, sizeof(resource));
-  if (strcmp(host, DNSSDHostName) == 0)
-    ptr = "localhost";
+#ifdef HAVE_DNSSD
+  if (DNSSDHostName)
+    nameptr = DNSSDHostName;
+  else
+#endif
+  if (ServerName)
+    nameptr = ServerName;
+  else
+    nameptr = NULL;
+
+  if (nameptr)
+  {
+    int host_len,
+        server_name_len;
+
+    httpSeparateURI(HTTP_URI_CODING_ALL, ptr,
+		    scheme, sizeof(scheme), userpass, sizeof(userpass), host,
+		    sizeof(host), &port, resource, sizeof(resource));
+
+    host_len = strlen(host);
+    if (strcmp(host + host_len - 6, ".local") == 0)
```

Comment:
Should we hardcode ``local.`` domain? Perhaps domain should be fetched somehow from [Avahi](http://www.avahi.org/doxygen/html/client_8h.html#aa6d5c36bc5fd18752b575c15427b52ae)

Reply:
We are comparing to Avahi's DNS-SD hostname, so we only care about .local.

Reply:
I doubt it is common, but mdns nss plugin can configure also different domains than .local. Also, can there be a difference between name ending with dot and without it? In DNS, trailing dot means that name does not get applied search domains from /etc/resolv.conf. Can something like that happen in Avahi? I think trailing dot could be squashed by:

```c
if (host_len>1 && host[host_len-1] == '.')
    host[--host_len] = '\0';
```